crates/control/src/visual_referee_filter.rs

```diff
@@ -0,0 +1,85 @@
+use num_traits::cast::FromPrimitive;
+use rand::prelude::*;
+use std::time::Duration;
+
+use color_eyre::{eyre::Context, Result};
+use context_attribute::context;
+use framework::MainOutput;
+use spl_network_messages::VisualRefereeDecision;
+use spl_network_messages::{PlayerNumber, VisualRefereeMessage};
+use types::{
+    hardware::Interface, messages::OutgoingMessage, CycleTime, FilteredWhistle, PrimaryState,
+};
+
+pub struct VisualRefereeFilter {
+    last_primary_state: PrimaryState,
+}
+
+#[context]
+pub struct CreationContext {
+    pub player_number: Parameter<PlayerNumber, "player_number">,
+}
+
+#[context]
+pub struct CycleContext {
+    pub primary_state: Input<PrimaryState, "primary_state">,
+    pub cycle_time: Input<CycleTime, "cycle_time">,
+    pub filtered_whistle: Input<FilteredWhistle, "filtered_whistle">,
+    pub player_number: Parameter<PlayerNumber, "player_number">,
+
+    pub hardware: HardwareInterface,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub last_primary_state: MainOutput<Option<PrimaryState>>,
+}
+
+impl VisualRefereeFilter {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            last_primary_state: PrimaryState::Unstiff,
+        })
+    }
+
+    pub fn cycle(&mut self, context: CycleContext<impl Interface>) -> Result<MainOutputs> {
+        let send_game_controller_vr_return_message = matches!(
+            (self.last_primary_state, *context.primary_state),
+            (PrimaryState::Set, PrimaryState::Playing)
+                | (
+                    PrimaryState::Playing,
+                    PrimaryState::Finished | PrimaryState::Ready
+                )
+        );
+        self.last_primary_state = *context.primary_state;
+
+        // A random visual referee decision
+        let mut rng = thread_rng();
+        let gesture = VisualRefereeDecision::from_u32(rng.gen_range(1..=13)).unwrap();
+
+        if send_game_controller_vr_return_message {
```

Comment:
Please don't abbreviate:

```suggestion
        if send_game_controller_visual_referee_return_message {
```

Maybe even `send_visual_referee_message`.

Reply:
OK. Will change it

Reply:
Thanks but since you also used the name "game controller visual referee message" in the networking code (for the name of the deduplicated function) I would favor `send_visual_referee_message`. This is accurate and short.

Reply:
Done